alibi/explainers/ale.py

```diff
@@ -0,0 +1,479 @@
+    sharey
+        A parameter specifying whether the y-axis of the ALE curves should be on the same scale
+        for several features. Possible values are 'all', 'row', None.
+    constant
+        A parameter specifying whether the constant zeroth order effects should be added to the
+        ALE first order effects.
+    ax
+        A matplotlib axes or a numpy array of matpotlib axes to plot on.
+    line_kw
+        Keyword arguments passed to the `plt.plot` function.
+    fig_kw
+        Keyword arguments passed to the `fig.set` function.
+
+    Returns
+    -------
+    An array of matplotlib axes with the resulting ALE plots.
+
+    """
+    import matplotlib.pyplot as plt
+    from matplotlib.gridspec import GridSpec
+
+    if features == 'all':
+        features = range(0, len(exp.feature_names))
+    else:
+        for ix, f in enumerate(features):
+            if isinstance(f, str):
+                try:
+                    f = np.argwhere(exp.feature_names == f).item()
+                except ValueError:
+                    raise ValueError(f"Feature name {f} does not exist.")
+            features[ix] = f
+    n_features = len(features)
+
+    if targets == 'all':
+        targets = range(0, len(exp.target_names))
+    else:
+        for ix, t in enumerate(targets):
+            if isinstance(t, str):
+                try:
+                    t = np.argwhere(exp.target_names == t).item()
+                except ValueError:
+                    raise ValueError(f"Target name {t} does not exist.")
+            targets[ix] = t
+
+    # make axes
+    if ax is None:
+        fig, ax = plt.subplots()
+
+    if isinstance(ax, plt.Axes) and n_features != 1:
+        ax.set_axis_off()  # treat passed axis as a canvas for subplots
+        fig = ax.figure
+        n_cols = min(n_cols, n_features)
+        n_rows = math.ceil(n_features / n_cols)
+
+        axes = np.empty((n_rows, n_cols), dtype=np.object)
+        axes_ravel = axes.ravel()
+        # gs = GridSpecFromSubplotSpec(n_rows, n_cols, subplot_spec=ax.get_subplotspec())
+        gs = GridSpec(n_rows, n_cols)
+        for i, spec in zip(range(n_features), gs):
+            # determine which y-axes should be shared
+            if sharey == 'all':
+                cond = i != 0
+            elif sharey == 'row':
+                cond = i % n_cols != 0
+            else:
+                cond = False
+
+            if cond:
+                axes_ravel[i] = fig.add_subplot(spec, sharey=axes_ravel[i - 1])
+                continue
+            axes_ravel[i] = fig.add_subplot(spec)
+
+    else:  # array-like
+        if isinstance(ax, plt.Axes):
+            ax = np.array(ax)
+        if ax.size < n_features:
+            raise ValueError(f"Expected ax to have {n_features} axes, got {ax.size}")
+        axes = np.atleast_2d(ax)
+        axes_ravel = axes.ravel()
+        fig = axes_ravel[0].figure
+
+    # make plots
+    if line_kw is None:
```

Comment:
maybe would be nice to have the defults and properties handling at the start of the fcn?